RecoEgamma/ElectronIdentification/src/ElectronMVAEstimatorRun2Fall17.cc

```diff
@@ -0,0 +1,428 @@
+  if( eleRecoPtr.get() == nullptr )
+    throw cms::Exception("MVA failure: ")
+      << " given particle is expected to be reco::GsfElectron or pat::Electron," << std::endl
+      << " but appears to be neither" << std::endl;
+
+  return fillMVAVariables(eleRecoPtr.get(), conversions, theBeamSpot.product(), theRho);
+}
+
+// A function that should work on both pat and reco objects
+std::vector<float> ElectronMVAEstimatorRun2Fall17::
+fillMVAVariables(const reco::GsfElectron* eleRecoPtr, const edm::Handle<reco::ConversionCollection> conversions, const reco::BeamSpot *theBeamSpot, const edm::Handle<double> rho) const {
+
+
+  // Both pat and reco particles have exactly the same accessors, so we use a reco ptr
+  // throughout the code, with a single exception as of this writing, handled separately below.
+  auto superCluster = eleRecoPtr->superCluster();
+
+  // Pure ECAL -> shower shapes
+  float see            = eleRecoPtr->full5x5_sigmaIetaIeta();
+  float spp            = eleRecoPtr->full5x5_sigmaIphiIphi();
+  float OneMinusE1x5E5x5 = 1. - eleRecoPtr->full5x5_e1x5() / eleRecoPtr->full5x5_e5x5();
+  float R9             = eleRecoPtr->full5x5_r9();
+  float etawidth       = superCluster->etaWidth();
+  float phiwidth       = superCluster->phiWidth();
+  float HoE            = eleRecoPtr->full5x5_hcalOverEcal(); //hadronicOverEm();
+  // Endcap only variables
+  float PreShowerOverRaw  = superCluster->preshowerEnergy() / superCluster->rawEnergy();
+
+  // To get to CTF track information in pat::Electron, we have to have the pointer
+  // to pat::Electron, it is not accessible from the pointer to reco::GsfElectron.
+  // This behavior is reported and is expected to change in the future (post-7.4.5 some time).
+  bool validKF= false;
+  reco::TrackRef myTrackRef = eleRecoPtr->closestCtfTrackRef();
+  const pat::Electron * elePatPtr = dynamic_cast<const pat::Electron *>(eleRecoPtr);
+  // Check if this is really a pat::Electron, and if yes, get the track ref from this new
+  // pointer instead
+  if( elePatPtr != nullptr )
+    myTrackRef = elePatPtr->closestCtfTrackRef();
+  validKF = (myTrackRef.isAvailable() && (myTrackRef.isNonnull()) );
+
+  //Pure tracking variables
+  float kfhits         = (validKF) ? myTrackRef->hitPattern().trackerLayersWithMeasurement() : -1. ;
+  float kfchi2          = (validKF) ? myTrackRef->normalizedChi2() : 0;
+  float gsfchi2         = eleRecoPtr->gsfTrack()->normalizedChi2();
+
+  // Energy matching
+  float fbrem           = eleRecoPtr->fbrem();
+
+  float gsfhits         = eleRecoPtr->gsfTrack()->hitPattern().trackerLayersWithMeasurement();
+  float expectedMissingInnerHits = eleRecoPtr->gsfTrack()
+    ->hitPattern().numberOfLostHits(reco::HitPattern::MISSING_INNER_HITS);
+
+  reco::ConversionRef conv_ref = ConversionTools::matchedConversion(*eleRecoPtr,
+                                    conversions,
+                                    theBeamSpot->position());
```

Comment:
it would be good to consistently use camelCase rather than mixing it and underscores